File: output/xCodeEval_remove_8000/C++/Go/49a2a9990a0401b418fed56b6815c229.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"sort"
	"strconv"
)

type Solver struct {
	N     int
	K     int
	A     [][]int64
	ASum  []int64
	Best  int64
}

func NewSolver(N, K int, A [][]int64) *Solver {
	return &Solver{
		N:    N,
		K:    K,
		A:    A,
		ASum: make([]int64, N),
		Best: 0,
	}
}

func (s *Solver) addItems(knapsack []int64, size int, sum int64) {
	for k := s.K - size; k >= 0; k-- {
		knapsack[k+size] = max(knapsack[k+size], knapsack[k]+sum)
	}
}

func (s *Solver) recurse(start, end int, knapsack []int64) {
	if start >= end {
		return
	}

	if end-start == 1 {
		sum := int64(0)
		s.Best = max(s.Best, knapsack[s.K])

		for prefix := 1; prefix <= len(s.A[start]); prefix++ {
			sum += s.A[start][prefix-1]
			s.Best = max(s.Best, sum+knapsack[s.K-prefix])
		}

		return
	}

	mid := (start + end) / 2
	state := make([]int64, s.K+1)
	copy(state, knapsack)

	for i := start; i < mid; i++ {
		s.addItems(state, len(s.A[i]), s.ASum[i])
	}

	s.recurse(mid, end, state)
	state = make([]int64, s.K+1)
	copy(state, knapsack)

	for i := mid; i < end; i++ {
		s.addItems(state, len(s.A[i]), s.ASum[i])
	}

	s.recurse(start, mid, state)
}

func (s *Solver) Solve() {
	s.Best = 0
	knapsack := make([]int64, s.K+1)
	s.recurse(0, s.N, knapsack)
	fmt.Println(s.Best)
}

func max(a, b int64) int64 {
	if a > b {
		return a
	}
	return b
}

func main() {
	s := bufio.NewScanner(os.Stdin)
	w := bufio.NewWriter(os.Stdout)
	defer w.Flush()

	var N, K int
	fmt.Fscan(s, &N, &K)

	A := make([][]int64, N)
	for i := 0; i < N; i++ {
		var t int
		fmt.Fscan(s, &t)
		A[i] = make([]int64, t)
		for j := 0; j < t; j++ {
			fmt.Fscan(s, &A[i][j])
		}
		sort.Slice(A[i], func(x, y int) bool {
			return A[i][x] > A[i][y]
		})
		if t > K {
			A[i] = A[i][:K]
		}
		var sum int64
		for _, a := range A[i] {
			sum += a
		}
		A[i] = append(A[i], sum)
	}

	solver := NewSolver(N, K, A)
	solver.Solve()
}

File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/Java/49a2a9990a0401b418fed56b6815c229.java
import java.io.*;
import java.util.*;

class DSum {
    static class Solver {
        private final BufferedReader in;
        private final PrintWriter out;

        public Solver(BufferedReader in, PrintWriter out) {
            this.in = in;
            this.out = out;
        }

        public void solve() throws IOException {
            int N, K;
            StringTokenizer st = new StringTokenizer(in.readLine());
            N = Integer.parseInt(st.nextToken());
            K = Integer.parseInt(st.nextToken());

            List<List<Long>> A = new ArrayList<>();
            List<Long> A_sum = new ArrayList<>();

            for (int i = 0; i < N; i++) {
                st = new StringTokenizer(in.readLine());
                int t = Integer.parseInt(st.nextToken());
                List<Long> list = new ArrayList<>();
                for (int j = 0; j < t; j++) {
                    list.add(Long.parseLong(st.nextToken()));
                }
                if (t > K) {
                    list = list.subList(0, K);
                }
                A.add(list);
                long sum = 0;
                for (long num : list) {
                    sum += num;
                }
                A_sum.add(sum);
            }

            long best = 0;
            List<Long> knapsack = new ArrayList<>(Collections.nCopies(K + 1, 0L));
            recurse(0, N, knapsack, A, A_sum, K);

            out.println(best);
        }

        private void add_item(List<Long> knapsack, int size, long sum) {
            for (int k = knapsack.size() - size - 1; k >= 0; k--) {
                knapsack.set(k + size, Math.max(knapsack.get(k + size), knapsack.get(k) + sum));
            }
        }

        private void recurse(int start, int end, List<Long> knapsack, List<List<Long>> A, List<Long> A_sum, int K) {
            if (start >= end) {
                return;
            }

            if (end - start == 1) {
                long sum = 0;
                long localBest = knapsack.get(K);

                for (int prefix = 1; prefix <= A.get(start).size(); prefix++) {
                    sum += A.get(start).get(prefix - 1);
                    localBest = Math.max(localBest, sum + knapsack.get(K - prefix));
                }

                best = Math.max(best, localBest);

                return;
            }

            int mid = (start + end) / 2;
            List<Long> state = new ArrayList<>(knapsack);

            for (int i = start; i < mid; i++) {
                add_item(state, A.get(i).size(), A_sum.get(i));
            }

            recurse(mid, end, state, A, A_sum, K);
            state = new ArrayList<>(knapsack);

            for (int i = mid; i < end; i++) {
                add_item(state, A.get(i).size(), A_sum.get(i));
            }

            recurse(start, mid, state, A, A_sum, K);
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter out = new PrintWriter(System.out);
        Solver solver = new Solver(in, out);
        solver.solve();
        out.close();
    }
}
